From eec4e8d062d985efd93e0f11feff179a0fd1ef3f Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Mon, 20 Jan 2025 14:42:25 +0100
Subject: [PATCH] usb: dwc2: update gadget state when suspended

Update the gadget state when the host suspends or resumes the bus. It
allow the correct suspended/configured (or state configured before the
suspend happens) to be reported in sysfs.

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: I60a518ae3f23d790766a91c5c9b019844da8d9cd
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/430483
Reviewed-by: Pankaj DEV <pankaj.dev@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/usb/dwc2/core.h      |  2 ++
 drivers/usb/dwc2/core_intr.c | 10 ++++++++++
 2 files changed, 12 insertions(+)

--- a/drivers/usb/dwc2/core.h
+++ b/drivers/usb/dwc2/core.h
@@ -873,6 +873,7 @@ struct dwc2_hregs_backup {
  * @hibernated:		True if core is hibernated
  * @in_ppd:		True if core is partial power down mode.
  * @bus_suspended:	True if bus is suspended
+ * @suspended_from:	The device state before the host has suspended the bus.
  * @reset_phy_on_wake:	Quirk saying that we should assert PHY reset on a
  *			remote wakeup.
  * @phy_off_for_suspend: Status of whether we turned the PHY off at suspend.
@@ -1071,6 +1072,7 @@ struct dwc2_hsotg {
 	unsigned int hibernated:1;
 	unsigned int in_ppd:1;
 	bool bus_suspended;
+	enum usb_device_state suspended_from;
 	unsigned int reset_phy_on_wake:1;
 	unsigned int need_phy_for_wake:1;
 	unsigned int phy_off_for_suspend:1;
--- a/drivers/usb/dwc2/core_intr.c
+++ b/drivers/usb/dwc2/core_intr.c
@@ -383,6 +383,7 @@ void dwc2_wakeup_from_lpm_l1(struct dwc2
 		}
 
 		/* Inform gadget to exit from L1 */
+		usb_gadget_set_state(&hsotg->gadget, hsotg->suspended_from);
 		call_gadget(hsotg, resume);
 		/* Change to L0 state */
 		hsotg->lx_state = DWC2_L0;
@@ -441,6 +442,7 @@ static void dwc2_handle_wakeup_detected_
 
 			/* Change to L0 state, when no_clock_gating == true */
 			hsotg->lx_state = DWC2_L0;
+			usb_gadget_set_state(&hsotg->gadget, hsotg->suspended_from);
 			call_gadget(hsotg, resume);
 		} else {
 			/* Change to L0 state */
@@ -585,6 +587,9 @@ static void dwc2_handle_usb_suspend_intr
 			 */
 			hsotg->lx_state = DWC2_L2;
 
+			hsotg->suspended_from = hsotg->gadget.state;
+			usb_gadget_set_state(&hsotg->gadget, USB_STATE_SUSPENDED);
+
 			/* Call gadget suspend callback */
 			call_gadget(hsotg, suspend);
 		}
@@ -661,6 +666,9 @@ static void dwc2_handle_lpm_intr(struct
 			dev_dbg(hsotg->dev,
 				"Core is in L1 sleep glpmcfg=%08x\n", glpmcfg);
 
+			hsotg->suspended_from = hsotg->gadget.state;
+			usb_gadget_set_state(&hsotg->gadget, USB_STATE_SUSPENDED);
+
 			/* Inform gadget that we are in L1 state */
 			call_gadget(hsotg, suspend);
 		}
@@ -812,6 +820,8 @@ static int dwc2_handle_gpwrdn_intr(struc
 				if (ret)
 					dev_err(hsotg->dev,
 						"exit hibernation failed.\n");
+
+				usb_gadget_set_state(&hsotg->gadget, hsotg->suspended_from);
 				call_gadget(hsotg, resume);
 			} else {
 				ret = dwc2_exit_hibernation(hsotg, 1, 0, 1);
